Source entity and source-validation package

arz-c opened the pull request:
The Source entity is serde deserializable and has matching fields to the Bibify API's JSON output. We are currently missing an implementation for measuring the credibility score of a given source. The source-validation package uses Bibify and serde_json to create a Source object given a website URL or book name.

arz-c committed:
create Source without credibility

arz-c committed:
create source validation package

barnaste reviewed (commented):
Looks great! The code in `source_validation/src/main.rs` looks clean, although I reckon it belongs in a different file, maybe even just `lib.rs` if you're not planning on introducing any other modules. At its core we're treating `source_validation` as a library crate.

The `Deserialize` implementation for `PublicationDate` looks neat! The logic is non-trivial though so having some tests should improve clarity and make sure things are working how we expect them to! I would prioritize this before we merge the branch into main.

Also, I'm wondering if there's a design decision behind choosing to make certain fields in `WebsiteInfo` and `BookInfo` optional. Is there ever a time where we won't know the `url` for a website, or the `title` for a book? From what I see in the `extract_...` functions those are the fields we're querying with, so I'm curious!

arz-c committed:
changed to lib crate, added tests

arz-c commented:
Good catch regarding `source_validation` being a library crate as opposed to a binary crate. I've renamed `source_validation/src/main.rs` to `source_validation/src/lib.rs` (I believe this is sufficient, but correct me if I'm wrong).

I moved the regex logic, previously found in `DateVisitor::visit_str`, to a new method `PublicationDate::parse_ymd_string`, and have added unit tests for the latter.

Regarding the abundance of optional fields in `WebsiteInfo` and `BookInfo`, I was thinking in terms of users manually filling in the fields (through the UI) of these high level entities. For example, when a user initially creates a blank source, I was imagining that it would be represented by a new `Source` object with every field set to `None`. If the user isn't directly interacting with the `Source` struct though, we can certainly remove the `Optional` overhead from some of the fields, as you mentioned.

arz-c committed:
remove some Optionals, add notes field
